presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PrefilterForLimitingAggregation.java

```diff
@@ -137,34 +138,36 @@ public PlanNode visitSort(SortNode sortNode, RewriteContext<Void> context)
         @Override
         public PlanNode visitLimit(LimitNode limitNode, RewriteContext<Void> context)
         {
-            PlanNode source = limitNode.getSource();
-            AggregationNode aggregationNode;
+            PlanNode source = rewriteWith(this, limitNode.getSource());
+            AggregationNode aggregationNode = null;
 
             if (source instanceof ProjectNode && ((ProjectNode) source).getSource() instanceof AggregationNode) {
                 aggregationNode = (AggregationNode) ((ProjectNode) source).getSource();
             }
             else if (source instanceof AggregationNode) {
                 aggregationNode = (AggregationNode) source;
             }
-            else {
-                return limitNode;
-            }
 
-            if (!aggregationNode.getGroupingKeys().isEmpty() && isScanFilterProject(aggregationNode.getSource())) {
+            if (aggregationNode != null &&
+                    !aggregationNode.getGroupingKeys().isEmpty() &&
+                    isScanFilterProject(aggregationNode.getSource()) &&
+                    !isGroupByOnlyPartitionKeys(aggregationNode, aggregationNode.getSource())) {
                 PlanNode rewrittenAggregation = addPrefilter(aggregationNode, limitNode.getCount());
-                if (rewrittenAggregation == aggregationNode) {
-                    return limitNode;
-                }
+                if (rewrittenAggregation != aggregationNode) {
+                    PlanNode newLimitNode;
+                    if (source == aggregationNode) {
+                        newLimitNode = replaceChildren(limitNode, ImmutableList.of(rewrittenAggregation));
```

Comment:
nit: this and the else only differ in the second argument, perhaps only compute that in the if/else and call replaceChildren only once:
```
PlanNode newSource;
if (source == aggregationNode) {
    newSource = ...
}
else {
...
}
PlanNode newLimitNode = replaceChildren(limitNode, ImmutableList.of(newSource));
```

Reply:
Better - just return with the right replace. Done